### PR TITLE
epicsEnvSet remove intential memory leak

### PR DESCRIPTION
**Pull request by mdavidsaver**
Address #137

### CODE REVIEW

**Comment by mdavidsaver** [comment]
This also get us down to three, actually distinct, implementations.

```
$ git ls-files|grep osdEnv
modules/libcom/src/osi/os/WIN32/osdEnv.c
modules/libcom/src/osi/os/default/osdEnv.c
modules/libcom/src/osi/os/vxWorks/osdEnv.c
```

File: modules/libcom/src/osi/os/WIN32/osdEnv.c
Comment (FreddieAkeroyd) [inline modules/libcom/src/osi/os/WIN32/osdEnv.c:46]: Do we need to check the return code,  or that value does not exceed `_MAX_ENV` ?

Comment (mdavidsaver) [reply modules/libcom/src/osi/os/WIN32/osdEnv.c:46]: Well, probably yes.  Although none of the other impls check for `ENOMEM` from `putenv()` or `setenv()`.

How should an error be communicated?  Change to a non-void return type?  (I hesitate to printf)

Comment (FreddieAkeroyd) [reply modules/libcom/src/osi/os/WIN32/osdEnv.c:46]: Changing the return type feels like a lot of work might potentially need to be done to trace the error through, though maybe that is the most correct.  Do you feel using `errPrintf` creates a dependency between the osi and other layers? It is probably unlikely for `_putenv_s` to fail, but if it did could we end up with a hard to trace error? That was probably more my concern, though errors caused by corrupt/missing environment variables are probably clearer than others to identify, so maybe it is not such an issue.

File: modules/libcom/src/osi/os/vxWorks/osdEnv.c
Comment (anjohnson) [inline modules/libcom/src/osi/os/vxWorks/osdEnv.c]: It would be nice to make `cp` a stack variable here, thus completely avoiding the need to call `malloc()` and `free()`. Since we're using gcc the following should be legal (but we still have to abide by the 'allocations before statements' rule so it must appear at the start of a block):

```C
    iocshEnvClear(name);
    {
        char cp[strlen (name) + strlen (value) + 2];

        strcpy (cp, name);
        strcat (cp, "=");
        strcat (cp, value);
        if (putenv(cp) < 0) {
```
The "Leaks memory" comment higher up could also be deleted since we won't any more.

File: modules/libcom/src/osi/os/WIN32/osdEnv.c
Comment (anjohnson) [reply modules/libcom/src/osi/os/WIN32/osdEnv.c:46]: The code Freddie is deleting was already calling `mallocMustSucceed()` which would suspend and dump a stack trace if it failed, and `errPrintf()` if the `putenv()` failed. Sticking with the latter wouldn't make matters any worse, so I say stick with that.

**Review by anjohnson — COMMENTED** [review]
A couple of suggestions.

**Comment by mdavidsaver** [comment]
I think I've addressed the comments.  Errors are now reported with `errlogPrintf()`, and calls to `cantProceed()` are removed.

On vxWorks, I've also made an (untested) attempt to use gcc on-stack dynamic arrays to avoid allocation for short strings.  The security conscious C programmer in me forces a fallback to `malloc()` for longer strings.

**Comment by mdavidsaver** [comment]
@anjohnson Do you plan to build/run test the vxWorks change prior to merge?  Otherwise, I will deal with any build fail which Jenkins turns up.

File: modules/libcom/src/osi/os/default/osdEnv.c
Comment (anjohnson) [inline modules/libcom/src/osi/os/default/osdEnv.c:34]: The whole point of our having the `osi/os/$(OS_CLASS)/` source structure was to avoid putting OS-specific conditionals into the source files since that makes them harder to read. The differences here are pretty minor so I understand the desire to not duplicate most of this code in an `osi/os/RTEMS/osdEnv.c` file, but they could be encapsulated even better like this:

```C
#ifdef __rtems__
#  define RTEMS_VERSION_INT VERSION_INT(__RTEMS_MAJOR__, __RTEMS_MINOR__, 0, 0)
#endif

#if defined(__rtems__) && RTEMS_VERSION_INT<VERSION_INT(4,10,0,0)
   /* newlib w/ RTEMS <=4.9 returns void */
#  define unsetenv_M(name) ({unsetenv(name); 0})
#else
#  define unsetenv_M(name) unsetenv(name)
#endif
```
Then remove the RTEMS-4.9 conditional below and call `unsetenv_M()` instead of `unsetenv()`.

Note that there is also a comment about `putenv()` immediately above the `epicsEnvUnset()` routine below which needs to be removed.

File: modules/libcom/src/osi/os/vxWorks/osdEnv.c
Comment (anjohnson) [inline modules/libcom/src/osi/os/vxWorks/osdEnv.c]: I would make the transition point much larger, say 512 bytes, which we would rarely if ever hit. There is enough stack for this kind of thing.

File: modules/libcom/test/epicsEnvTest.c
Comment (anjohnson) [inline modules/libcom/test/epicsEnvTest.c:77]: Why does the code below sometimes use `name` and sometimes a literal `"testChangeEnv"`? It's a little confusing to read, and up to the compiler whether all the literals would return the same pointer.

**Review by anjohnson — COMMENTED** [review]
I could build it on VxWorks but I don't have the ability to run any tests at the moment, my VME systems at work are turned off (except for the MVME2500 for Brendan to test RTEMS on). I suggest we wait until you do the merge to check the VxWorks build.

I added some additional comments here with some more code improvements though, sorry!

File: modules/libcom/test/epicsEnvTest.c
Comment (mdavidsaver) [reply modules/libcom/test/epicsEnvTest.c:77]: A (perhaps futile) attempt to detect if `epicsSetEnv()` stores, and later clobbers, the `const char*` passed to it.

File: modules/libcom/src/osi/os/vxWorks/osdEnv.c
Comment (mdavidsaver) [reply modules/libcom/src/osi/os/vxWorks/osdEnv.c]: Ok, your call.

File: modules/libcom/src/osi/os/default/osdEnv.c
Comment (mdavidsaver) [reply modules/libcom/src/osi/os/default/osdEnv.c:34]: Right, I hardly ever use the `({...})` construct, but I guess it should be compatible with old enough gcc and clang.  Would it work, and does it need to work, with suncc?

Digression...

I was thinking about putting the (conditional) definition of `RTEMS_VERSION_INT` in `epicsVersion.h`.  Maybe also `GCC_VERSION_INT`.  We use these somewhat widely, and the definitions should always be the same.

Comment (anjohnson) [reply modules/libcom/src/osi/os/default/osdEnv.c:34]: The `({ ... })` construct is only needed when building for RTEMS <= 4.9, the other branch of my suggestion doesn't need/use it. Are you asking whether RTEMS 4.9 was built with suncc? 😉 

Exporting `RTEMS_VERSION_INT` might have a deleterious effect on external modules which also define it and would not expect it to be already defined. I'm not objecting to the idea, but it might have some repercussions downstream. There is only one place each where we currently define and use that and `VXWORKS_VERSION_INT` although I haven't checked the RTEMS-5 branch.

Comment (mdavidsaver) [reply modules/libcom/src/osi/os/default/osdEnv.c:34]: > ... Are you asking whether RTEMS 4.9 was built with suncc? wink

Brain fart...

> Exporting `RTEMS_VERSION_INT` might have a deleterious effect ...

True, although at worst this would force adding another `#ifdef` (which I would include in `epicsVersion.h` as well).  I see both the benefits and costs as minor.  So I may add these definitions the next time it would be convenient, or not.

**Comment by mdavidsaver** [comment]
I've addressed the round 2 comments.

**Review by anjohnson — APPROVED** [review]
Thanks!

**Comment by mdavidsaver** [comment]
Merged as of 3eeebb74cd83757756acb2024560e602ee8f9118

**Pull request closed without merging**
